File: 0897-increasing-order-search-tree/0897-increasing-order-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def increasingBST(self, root: TreeNode) -> TreeNode:
        
        node = root
        stack = []
        pre = None
        smallest = None
        while stack or node:
            if node:
                stack.append(node)
                node = node.left
            else:    
                node = stack.pop()
                if not smallest:
                    smallest = node
                node.left = None
                if pre:
                    pre.right = node
                pre = node
                node = node.right
        return smallest

File: 0897-increasing-order-search-tree/0897-increasing-order-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode increasingBST(TreeNode root) {
        
        TreeNode node = root;
        Stack<TreeNode> stack = new Stack<>();
        TreeNode pre = null;
        TreeNode smallest = null;
        
        while (!stack.isEmpty() || node != null) {
            if (node != null) {
                stack.push(node);
                node = node.left;
            } else {
                node = stack.pop();
                if (smallest == null) {
                    smallest = node;
                }
                node.left = null;
                if (pre != null) {
                    pre.right = node;
                }
                pre = node;
                node = node.right;
            }
        }
        
        return smallest;
    }
}
